Move Git history view to scm-extra package

westbury opened the pull request:
#### What it does
The SCM view currently supports different SCM types through VSCode APIs.  However the History view supports Git only.  This PR refactors the history view, splitting apart the Git specific code from the history view.  This allows a single history view to be shown that tracks the selected repository regardless of whether the repository is a Git repository or a repository from another source manager type such as Mercurial.

This PR moves GitHistoryWidget to ScmHistoryWidget in the SCM package.  GitNavigableListWidget and other dependencies have also been moved.

The VSCode does not have API to support history.  Therefore we define our own API.  I could have implemented this in a similar way to the AmendSupport interface, where the provider provides an implementation.  However I decided instead to create a child container for each SCM type, so one child container for Git, one for Mercurial etc.  The history support implementations are bound in these containers.  The containers containing the repository options are now child containers to these.  This approach gives flexibility when adding extra SCM types, or extra an extra interface, or an implementation of an interface for a particular SCM type.  So, for example, someone could add support for SVN but not implement the History interface.  Someone else could add an extension that implements the History Support interface for SVN and bind it in the SVN child container.
  
#### How to test
This PR should be tested to ensure that it works with Git repositories exactly as before.  The only differences one should notice is that the view label has been changed from 'SCM' and 'Git History' to 'Source Control' and 'Source History'.  The label change was preferred by our UX team but if there are reasons for using the same label as VSCode then I can change it back and we will override it in our application.

#### Review checklist

- [x ] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

akosyakov commented:
Does it make sense to move Git History to SCM? There are already VS Code extensions providing git history. They should be used. cc @svenefftinge

westbury commented:
> There are already VS Code extensions providing git history. They should be used

Are you saying the current Git History should be removed completely from the Theia code base in favor of using VS Code extensions?

Unless I am missing something, the VS Code extensions are specific to a single SCM type.  Extensions written for Git won't work with Mercurial etc.  Our requirement is that the history view shows history for different SCM types, switching between a Git and Mercurial repository in the same way that the current SCM view switches seamlessly between Git and Mercurial repositories.

akosyakov commented:
> Are you saying the current Git History should be removed completely from the Theia code base in favor of using VS Code extensions?

Not remove, I don't think it makes sense to have it in the SCM extension, since each vscode extension will bring own history implementation.

> Unless I am missing something, the VS Code extensions are specific to a single SCM type. Extensions written for Git won't work with Mercurial etc. Our requirement is that the history view shows history for different SCM types, switching between a Git and Mercurial repository in the same way that the current SCM view switches seamlessly between Git and Mercurial repositories.

Yes, but each VS Code extension can provide a tree view or a webview to implement history specific to a SCM type.

I would prefer that we discuss it on dev meeting when Sven is back. Unfortunately in 2 weeks after EclipseCon. It is not clear to me how we should move forward.

akosyakov commented:
@westbury I've chatted with @svenefftinge about it. Let's keep `@theia/scm` with minimal code required both for Theia and VS Code extensions, but add a new package `@theia/scm-extra` which  contains reusable widgets for Theia extension. Would it work for you?

Also wonder whether you are interested into using `@theia/git` if the rest of community moves to use VS Code extension? The question is whether we can delete it after deprecation or we have to move it to own repo and someone has to maintain it.

akosyakov commented:
Looking at the PR: We should test that the layout with git history view created by old Theia is still can be opened. Theia has the layout migration contribution: https://github.com/eclipse-theia/theia/blob/f255f5a8cb22010aacb8196c8a8f248e111e9945/packages/core/src/browser/shell/shell-layout-restorer.ts#L88 We should add a new one if the layout restoration breaks. Please look at other implementations as examples.

westbury commented:
> add a new package `@theia/scm-extra` which contains reusable widgets for Theia extensions. Would it work for you?

Yes, moving into scm-extra would certainly work for us.  It makes sense that the scm package is the minimal needed to support the vscode SCM extensions.  The 'amend' support could also be moved, perhaps by allowing Theia extensions to contribute sections to be bottom of the SCM view.

We will obviously need some discussions here on what we do if the rest of the community abandon the Theia git extension.  Note the following concerns:

-  vscode extensions do not have full coverage.  For example there is no Mercurial extension that has anything equivalent to the history view, and this would be a loss of feature for us because Mbed Studio currently shows Mercurial history in Theia's history view by using the very flexible Inversify framework.

- vscode extensions do not provide a consistent UI.  Most of our users are not avid Mercurial users.  If they are using a library from a Mercurial repository then they would expect to be able to access and maintain it through the same UI as Git as far as possible.  vscode does this very well for the SCM view but then leaves each SCM extension to itself for anything outside the SCM view.

- The Inversify framework is very powerful.  It would be a big loss to abandon that and move to vscode extensions.  Extenders will no longer be able to rebind to customize UI provided by a vscode extension.

cc: @mcgordonite @thegecko @arekzaluski

westbury commented:
> I would prefer that we discuss it on dev meeting

Yes, let's discuss it tomorrow.  I'll be on the dev meeting.

akosyakov commented:
ok, added to the agenda:  https://github.com/eclipse-theia/theia/wiki/Dev-Meetings#agenda-2019-10-29

benoitf commented:
on Red Hat side, we're using only VS Code scm (we've removed all extensions from 'our' Theia that have VS Code extensions counterpart)

westbury commented:
I have now
- moved it to a new scm-extra package
- added an upgrader so the history view is restored in the layout
- rebased on top of the theming changes

Unfortunately there is now a failure in the integration test.  This fails for me locally even on master.  Many of the views seems to randomly fail to open and close.

westbury commented:
The integration tests now pass on CI again, so this is now ready for review.

@akosyakov We appreciate your considering this PR because being limited to what we get from VS Code extensions really does not work for us.  I understand this may move to a separate repository if you remove native Git extension.

akosyakov commented:
@westbury i will have a look next day, sorry for the delay

akosyakov commented on packages/scm/src/browser/scm-input.ts:
```diff
@@ -32,6 +33,7 @@ export interface ScmInputValidator {
 export interface ScmInputOptions {
     placeholder?: string
     validator?: ScmInputValidator
+    providerContainer?: interfaces.Container
```
exposing container is not good, why injection cannot be used?

westbury replied:
I do not know of a better way of supporting the requirements stated in the third paragraph in the initial PR comments.  To give flexibility going forward, I want to support the following:

- an extension package can add support for the 'history' api when the package contributing the scm did not provide a 'history' api.
- an extension package can define a new api and be a consumer of that api.  It can implement the api for specific SCMs, or downstream packages can implement the API for a SCM.

There must be a way of getting the container for the SCM type, otherwise packages can't add implementations to it.  Neither the container nor the history implementation can be injected into the ScmHistoryWidget because it will change as the user switches between repositories of different SCM types.  Let me know if you can think of a better way of doing this.
 
  three

akosyakov replied:
There can be something like:
```ts
interface ScmRepositoryFactory {
    readonly id: string;
    get(option: ScmRepositoryOptions): ScmRepository
}
```

get method will create a child container, inject ScmRepositoryOptions, different extensions can contribute different factories

not sure about id, it probably cannot be scm provider id, but should be more something like `git` or `plugin`, `plugin` will handle all scm providers contributed by VS Code extensions

westbury replied:
But it's not about getting the ScmRepository, it's about getting the specific API implementations.  Eg. an implementation of ScmHistorySupport, or perhaps the implementation to support the 'Amend' button.  So the 'get' function would have to be:
```ts
    get<T>(option: ScmRepositoryOptions, serviceIdentifier: interfaces.ServiceIdentifier<T>): T | undefined;
```
and then the implementation would have to use the id to find the container that has the bindings appropriate for the SCM type.  The view, when notified of a new selected ScmRepository, would have to get the id from the ScmRepoistory, then that would be looked up in a map to find the container with the bindings for that SCM type (or as ServiceIdentifier can be string as well as symbol, use the id to bind the container).  That would mean looking up the scm type container by id every time.  Then it would have to create a grandchild container (a child of the scm type container to contain the ScmRepositoryOptions), but it would have to recreate the grandchild container for each API.  This seems overly complex.

I could change it so we pass a callback to get API implementations, instead of passing the container.

westbury replied:
@akosyakov  I've pushed https://github.com/eclipse-theia/theia/pull/6381/commits/5bedb57f988aa789fc9339f9c23851aba8c1004c so that now the factory is injected.  Let me know if this is ok.

akosyakov commented on packages/scm/src/browser/scm-provider.ts:
```diff
@@ -69,11 +69,28 @@ export interface ScmCommand {
+}
+
+export interface ScmFileChange {
+    uri: string;
```
`readonly`?

westbury replied:
I've changed this and some others to readonly.

akosyakov commented on packages/scm/src/browser/scm-provider.ts:
```diff
@@ -69,11 +69,28 @@ export interface ScmCommand {
+    authorTimestamp: string;
+    authorDateRelative: string;
+    fileChanges: ScmFileChange[];
+    commitDetailOptions: {};
```
It is not used in scm module? Could we get rid of unused properties and move them downstream extensions which are using it?

westbury replied:
You are correct this is not read in the scm package.  However it is read in both the git and the scm-extra packages.  In scm-extra package it is passed through to the 'open' method, so is used in git-commit-detail-open-handler.ts, doOpen.  I don't see how it can be moved down to the git and scm-extra packages because it must be the same value in both cases.

akosyakov replied:
Cannot scm-extra define some additional types with `is` function for down-casting and git extension making use of them?

akosyakov commented on packages/scm-extra/src/browser/history/scm-history-widget.tsx:
```diff
@@ -15,73 +15,91 @@
+        this.title.caption = SCM_HISTORY_LABEL;
         this.title.iconClass = 'fa git-history-tab-icon';
         this.title.closable = true;
         this.addClass('theia-git');
```
Should not we get rid of `git` notion in the scm extra?

westbury replied:
I have made the changes to the css.  I've also added another sentence to the change log for the breaking css changes.

akosyakov commented on packages/git/src/browser/git-scm-provider.ts:
```diff
@@ -362,21 +418,22 @@ export namespace GitScmProvider {
-    export const Factory = Symbol('GitScmProvider.Factory');
-    export type Factory = (options: GitScmProviderOptions) => GitScmProvider;
-    export function createFactory(ctx: interfaces.Context): Factory {
+    export type ContainerFactory = (options: GitScmProviderOptions) => interfaces.Container;
```
we should use injection not passing a container around

akosyakov commented on packages/scm/src/browser/style/index.css:
```diff
@@ -173,48 +173,49 @@
 }
 
-.theia-scm:focus-within .scmItem.theia-mod-selected {
+*:focus-within .scmItem.theia-mod-selected {
```
It looks bogus,
- top-level classes should follow our coding guidlines and start with `theia-` prefix
- another element is focused within body element and scm item is highlighted as selected?

westbury replied:
OK, I've fixed that by having both theia-scm and theia-git classes set on the git widgets.

akosyakov commented on packages/scm/src/browser/scm-provider.ts:
```diff
@@ -69,11 +69,23 @@ export interface ScmCommand {
+    readonly authorEmail: string;
+    readonly authorDateRelative: string;
+    readonly fileChanges: ScmFileChange[];
+    readonly commitDetailOptions: {};
```
`scm-extra` and `git` extension are very likely going to be moved to a separate repo soon and replace by vscode `git` and `git-ui` extensions.

There should be a way that `git` and `scm-extra` can add new properties without adding them to `scm` package. `scm` package should contain only common part between features needed to support VS Code extensions and scm-extra but not more.

westbury replied:
I've moved fileChanges and commitDetailOptions out of scm package, also ScmFileChange.  ScmCommit remains for time being as it is used by 'amend' support.

westbury commented:
@akosyakov I've addressed the remaining two issues:
- moved fileChanges, commitDetailOptions and ScmFileChange into scm-extra
- replaced container injection with factory injection

Let me know if you see anything else.

akosyakov commented:
@westbury sorry, we are preparing Gitpod release and i was busy, i would try to have a look today or tomorrow

akosyakov commented on packages/scm/src/browser/scm-provider.ts:
```diff
@@ -16,9 +16,15 @@
 import URI from '@theia/core/lib/common/uri';
 
+export const ScmFactory = Symbol('ScmFactory');
+export interface ScmFactory {
```
it is not used in scm

akosyakov commented on packages/scm/src/browser/scm-provider.ts:
```diff
@@ -69,11 +76,12 @@ export interface ScmCommand {
-    authorEmail: string,
-    authorDateRelative: string
+    readonly id: string;  // eg Git sha or Mercurial revision number
+    readonly commitDetailUri: URI;
```
this property is not used in scm

akosyakov commented on packages/scm-extra/src/browser/history/scm-history-widget.tsx:
```diff
@@ -15,92 +15,158 @@
+
+        const repository = this.scmService.selectedRepository;
+        if (repository) {
+            this.historySupport = repository.provider.get && repository.provider.get(ScmHistorySupport);
```
Why cannot we introduce a new interface like:
```ts
interface ScmHistoryProvider extends ScmProvider {
    historySupport?: ScmHistorySupport
}
namespace ScmHistoryProvider {
    export function is(scmProvider: ScmProvider | undefined): provider is ScmHistoryProvider {
        retunrn scmProvider && 'historySupport' in scmProvider;
    }
}
```

so here we just do down cast check without passing containers around and providing unused by scm get method to lookup random service?

westbury replied:
That won't work because two different packages might both be providing a custom API implementation to an SCM.  Currently ScmHistorySupport is the only extension API.  But suppose there was another extension API.  For example, suppose we decide to move the Amend support out of the scm package into yet another package.  Or suppose an extender consumes another custom API.  The provider can't implement all those interfaces (ScmHistoryProvider, ScmAmendProvider, etc) when the definitions of those interfaces come from different packages that may or may not be installed.

This is support for binding with a scope of SCM type.  It allows Alice to define an interface with the functions she needs to implement her SCM view.  It allows Bob to bind an implementation of that interface to the Git container.  It allows Chris to bind an implementation of that interface to the SVN container.  All are doing so in their own packages (not changing scm or scm-extra packages).

westbury replied:
On thinking more about this, I may be able to come up with something better.  It will be tricky to inject implementations into the SCM type child container from a different package than the one that added the SCM type.  I may push some changes tomorrow.

akosyakov replied:
> The provider can't implement all those interfaces (ScmHistoryProvider, ScmAmendProvider, etc) when the definitions of those interfaces come from different packages that may or may not be installed.

Should not provider depends on these packages if he is going to implement them? I think it is what we have now, git extension depends on scm and scm-extra and can implement both interfaces, it will depends on scm-amend if it is required too.

if a provider only interested in type he has to add these packages as dev dependencies and import only types. In this case scm-extra and scm-amend don't come with git extension by default.

westbury replied:
> Should not provider depends on these packages if he is going to implement them? I think it is what we have now, git extension depends on scm and scm-extra and can implement both interfaces, it will depends on scm-amend if it is required too.

I anticipated that it will be useful to add a new API and add a Git implementation without needing to modify the git package.  However that is not essential for us so I made the changes to simply get the history support directly from the provider.

The problem is, however, that the provider cannot depend on the history support.  It cannot have history support injected into it because the provider is injected into the history support.  Furthermore git-scm-provider.ts cannot have anything imported from git-history-support because that would also be a circular dependency.  Removing the provider as a dependency to git-history-support would not be good enough because git-history-support depends on GitRepositoryTracker which depends on GitRepositoryProvider.  So I was right back to removing all the history imports from git-scm-provider.

ScmFactory has been removed.  The scm-type containers have been removed and scm-input has been reverted.  The child container bindings have been moved out of git-scm-provider so it can bind history stuff.  You will see in that binding code a small hack to set the history support into the provider.

akosyakov replied:
ok, thanks, could we remove `ScmProvider.get` please?

akosyakov commented on packages/scm/src/browser/scm-provider.ts:
```diff
@@ -33,6 +39,7 @@ export interface ScmProvider extends Disposable {
     readonly onDidChangeStatusBarCommands?: Event<ScmCommand[] | undefined>;
 
     readonly amendSupport?: ScmAmendSupport;
+    readonly get?: <T>(id: interfaces.ServiceIdentifier<T>) => T | undefined;
```
it is not used by scm, it could be replaced by more specialized intereface and down casting with is operator

akosyakov replied:
that should be removed

akosyakov commented:
Could someone from your team do the functional testing that everything works as before for changed parts? It would help a lot.

What do we do with scm-amend? I guess we will need cooperate somehow that it works for VS Code and Theia extensions. Could file an issue please?

westbury commented:
> Could someone from your team do the functional testing that everything works as before for changed parts? It would help a lot.

@federicobozzini has been volunteered to do this testing.  He will be back from vacation on Thursday.

federicobozzini commented:
I've checked some of the most basic functionalities (init, commit, amend, stash, unstash, ...) and I've found only one difference so far.

How to reproduce it:

In  an empty workspace add a folder `a` with a subfolder `b`. Initialize `b` to be a git repository. `b` is now the selected repository of the workspace (as you can see from the left-bottom selector). Now initialize `a` to be a repository. If you restart Theia, on master `b` remains the default repository, while on @westbury branch `a` is selected.

akosyakov commented:
We should check that git history, diff works the same way. It's the code which was moved.

federicobozzini commented:
I'm not sure why I'm seeing this difference in behavior, maybe just a cache issue.

There also seems an issue with the git history.

In some cases when in the Git history panel a commit is expanded and a file is clicked to see its changes, it the file is opened without showing its name in the monaco navigator.

![Screenshot from 2020-02-12 17-46-58](https://user-images.githubusercontent.com/8626576/74361989-c0fb2c00-4dbf-11ea-9e08-1c317706599d.png)

westbury commented:
Thanks @federicobozzini for testing this.  I don't think either of the two issues are problems on the branch.

> In an empty workspace add a folder `a` with a subfolder `b`. Initialize `b` to be a git repository. `b` is now the selected repository of the workspace (as you can see from the left-bottom selector). Now initialize `a` to be a repository. If you restart Theia, on master `b` remains the default repository, while on @westbury branch `a` is selected.

I can't reproduce this.  If I stop Theia while b is the selected repository, then git init a, then restart Theia, I see b selected but a now added to the list.  Same if I don't stop Theia.  It could be a cache issue as you say if somehow the selection was cleared so the outer repository would be selected as the first in the list.

> In some cases when in the Git history panel a commit is expanded and a file is clicked to see its changes, it the file is opened without showing its name in the monaco navigator.

This one is caused by the label provider.  It can be fixed by changing the following line to return 'undefined' (now that that is allowed courtesy of 11be0c0ad25ab3946fe28f0dc9d2e9e48691d34a)

https://github.com/eclipse-theia/theia/blob/46a2d510b9aa3407d2a4ef3af0f403beeaebc204/examples/api-samples/src/browser/sample-dynamic-label-provider-contribution.ts#L78

However the fix to the label provider is bigger than that so needs to happen under a separate PR.  The bug is on the master branch.

akosyakov commented:
@federicobozzini but generally It looks good to you?

federicobozzini commented:
I've done another round of testing and I've found only one minor issue. When focusing on an element in the history panel, a blue border appears around some elements in the history.

![Screenshot from 2020-02-20 10-01-46](https://user-images.githubusercontent.com/8626576/74923927-826dfe80-53c9-11ea-908e-c9fe0dca2c1c.png)

It looks like it's caused by some css rules that are still applied using the `.theia-git` selector, while instead the new `.theia-scm` should be used.

Except for that, I can't see any meaningful difference.

akosyakov commented:
thanks, blue border should be fixed before merging

I test now too.

thegecko commented:
> thanks, blue border should be fixed before merging

Agreed, this will be addressed on Monday

akosyakov commented:
Please have a look at https://github.com/eclipse-theia/theia/pull/6381#discussion_r382606553 too. This API is not used in the scm extension.

I did some testing and could not find anything completely broken. Only thing which bothered me how we name `Git History` now. It is `Source History`, maybe `File History` or `File Timeline` is better?  cc @svenefftinge 
<img width="399" alt="Screenshot 2020-02-21 at 15 27 21" src="https://user-images.githubusercontent.com/3082655/75042965-92680a00-54bf-11ea-8f88-8c1765ab6f9a.png">
<img width="335" alt="Screenshot 2020-02-21 at 15 26 52" src="https://user-images.githubusercontent.com/3082655/75042970-93993700-54bf-11ea-8c77-3879298d54a6.png">

@jbicker @kittaakos maybe you can test a bit too, git view, diff and history in particularly?

akosyakov commented:
Discussed with @svenefftinge to name it `History` without `Source`.

westbury commented:
I've made the following changes:
1. 'Source History' -> 'History'
2. The blue outline has been removed.  The focus css was moved from theia-git to theia-scm.  I have checked it is not needed in theia-git so it is a move not a copy.
3 Removed the extraneous 'get' function declaration from ScmProvider.
I believe that addresses all the outstanding comments.

thegecko commented:
Thanks @westbury 

@federicobozzini  could you re-test?

federicobozzini commented:
Re-tested. The issue I pointed out is not present anymore.

akosyakov commented:
@thegecko @westbury Is it very urgent? Could we await release and land it on Friday? @marcdumais-work release is this Thursday correct?

marcdumais-work commented:
> release is this Thursday correct?

Correct

thegecko commented:
> Could we await release and land it on Friday?

No rush, don't want to risk de-stabilising before a release  :)

westbury commented:
I have rebased onto the 0.16.0 release.  There were no conflicts other than the usual changelog and version numbers.  However there is a test failure which appeared.  I have seen this test failure before but it went away after a rebase.  It is intermittent in that it appears when rebasing then goes away with the next rebase.

I have fixed the tests by making a couple of changes in editor-decoration-style.ts and git-repository-provider.spec.ts.  The offending code in editor-decoration-style is never executed when the branch is in a passing state.

vince-fugnitto commented:
In the future (once it is stabilized), we might benefit from using the new **timeline** api with the contributed git commit history view:

https://github.com/microsoft/vscode-docs/blob/ba22868d9ac135dd47265d8f6c2e6d160a2b142a/release-notes/v1_42.md#timeline-view

![image](https://user-images.githubusercontent.com/40359487/76314157-a926a300-62ac-11ea-9335-28f5fd4a02ed.png)

akosyakov commented:
@vince-fugnitto could you file an issue we have to implement a widget in the scm extension which supports such API

vince-fugnitto commented:
> @vince-fugnitto could you file an issue we have to implement a widget in the scm extension which supports such API

I believe it is a proposed API still, and looking at the view it seems to be for the `Explorer`?

akosyakov commented:
@vince-fugnitto agreed maybe we have a look at it in half a year
